### ISSUE TITLE
Calling too many methods performing REST requests to Discord in a parallel context causes the WebSocket to irrecoverably disconnect on all shards

### ISSUE THREAD

**Issue by Alex1304** [Discord4J/Discord4J#513]
**To Reproduce:** 

This issue happens when many methods involving REST requests are called in parallel. You can easily reproduce this issue by followings these steps:

1. Create a `Flux` that generates thousands of random invalid `Snowflake` objects. 
2. `flatMap` them to `Channel` objects (with an `onErrorResume` to ignore errors due to channels not existing)

Code snippet:
```Java
// Put this code in the body of a bot command (e.g !test) 
Flux.<Snowflake>create(emitter -> {
    for (long l = 1 ; l < 100_000 ; l++)
        emitter.next(Snowflake.of(l));
}).flatMap(snowflake -> discordClient.getChannelById(snowflake)
        .onErrorResume(e -> Mono.empty()))
.subscribe();
```
**The bug doesn't appear immediately upon executing this code. You should let this code run for 30 seconds - 1 minute minimum in order to see it.**

Of course, this is just for testing purposes in order to reproduce the bug, but this is not really a valid use case. That's why I would like to explain my own use case and why this issue is particularly irritating for me. Personally I work on a bot which core feature is to send news messages to servers. Server owners just need to indicate the bot where to send the news messages, and that's it. Channel IDs are stored in a relational database. When the bot is ready to broadcast a piece of news, it will first gather the IDs of all configured channels across all servers, and then starts sending the message in every single channel. So in my case, the code looks more like this (simplified here of course):

```Java
Consumer<MessageCreateSpec> messageToBroadcast; // the message to send to all channels
Flux<Snowflake> channelIds = getChannelIdsFromDatabase();

channelIds.flatMap(id -> discordClient.getChannelById(id)
        .ofType(MessageChannel.class)
        .onErrorResume(e -> Mono.empty())
.flatMap(channel -> channel.createMessage(messageToBroadcast)
        .onErrorResume(e -> Mono.empty())
.subscribe();
```

Since my bot is in more than 4000 servers, this Flux of Snowflakes actually contains a LOT of channel IDs!
Only difference with the first example is that the snowflakes are actually valid, so most of the Channels are gathered from the Store and it doesn't hit REST. Well, a few of them will still hit REST since it may happen that the database contains channelIds that reference to channels that are either deleted or inaccessible, but the issue in my case actually happens in the `createMessage`, which is effective hitting REST on each call. 

**Expected Behavior:** 

The concurrency on REST requests should be properly handled so that it doesn't crash the websocket

**Actual Behavior:** 

After a few seconds running, the websocket will end up closing with reason code 1000, resulting in all shards to disconnect one by one:

```
2019-04-10 00:00:36.340 [reactor-http-epoll-8] INFO  discord4j.gateway.inbound.1 - Receiver completed on shard 1
2019-04-10 00:00:36.340 [reactor-http-epoll-8] INFO  discord4j.gateway.inbound.1 - Forwarding close reason: 1000
2019-04-10 00:00:36.341 [reactor-http-epoll-8] ERROR discord4j.gateway.client.1 - Gateway client error: discord4j.common.close.CloseException: WebSocket closed: 1000
2019-04-10 00:00:36.341 [reactor-http-epoll-8] INFO  discord4j.gateway.client.1 - Retry attempt 1 in 2660 ms
2019-04-10 00:00:42.467 [reactor-http-epoll-2] INFO  discord4j.gateway.inbound.2 - Receiver completed on shard 2
2019-04-10 00:00:42.467 [reactor-http-epoll-2] INFO  discord4j.gateway.inbound.2 - Forwarding close reason: 1000
2019-04-10 00:00:42.468 [reactor-http-epoll-2] ERROR discord4j.gateway.client.2 - Gateway client error: discord4j.common.close.CloseException: WebSocket closed: 1000
2019-04-10 00:00:42.468 [reactor-http-epoll-2] INFO  discord4j.gateway.client.2 - Retry attempt 1 in 2413 ms
```

Despite the logs saying "Retry attempt in xxx", the websocket will never reattempt connection and the bot will go offline. 

**Version:** 

Discord4J v3.0.1

**Other:** 

In the meantime this bug gets fixed (hopefully), I found some workarounds in order to make the code work. The following alternatives work and do not produce the error:

Workaround 1: delaying each Snowflake (awful):
```Java
Flux.<Snowflake>create(emitter -> {
    for (long l = 1 ; l < 100_000 ; l++)
        emitter.next(Snowflake.of(l));
}).delayElements(Duration.ofMillis(200))
.flatMap(snowflake -> discordClient.getChannelById(snowflake)
        .onErrorResume(e -> Mono.empty()))
.subscribe();
```

Workaround 2: using `concatMap` instead of `flatMap` (ensures that all requests are done one by one, works great but hella slow cuz no parallelism):
```Java
Flux.<Snowflake>create(emitter -> {
    for (long l = 1 ; l < 100_000 ; l++)
        emitter.next(Snowflake.of(l));
}).concatMap(snowflake -> discordClient.getChannelById(snowflake)
        .onErrorResume(e -> Mono.empty()))
.subscribe();
```

Workaround 3: using the second argument of `flatMap` to limit concurrency (the best option I could find so far):
```Java
Flux.<Snowflake>create(emitter -> {
    for (long l = 1 ; l < 100_000 ; l++)
        emitter.next(Snowflake.of(l));
}).flatMap(snowflake -> discordClient.getChannelById(snowflake)
        .onErrorResume(e -> Mono.empty()), Runtime.getRuntime().availableProcessors())
.subscribe();
```
So yeah this issue seems to have something to do with concurrency.

**Comment by quanticc** [comment]
Thanks for the detailed report. We have located the root cause and will be pushing a fix soon.

**Comment by quanticc** [comment]
Can you test `474fcee` from jitpack: https://github.com/Discord4J/Discord4J/wiki/Using-Jitpack

**Comment by Alex1304** [comment]
Sure thing. Will test this later today and let you know if I still experience the issue or not

**Comment by Alex1304** [comment]
Okay so I have things to say about this update.

First of all, here is the code I tested it with. It's basically the same as before, except that I added an instruction to print in logs:

```Java
Flux.<Snowflake>create(emitter -> {
	for (long l = 1 ; l < 100_000 ; l++)
	 	 emitter.next(Snowflake.of(l));
}).concatMap(snowflake -> discordClient.getChannelById(snowflake)
	 .onErrorResume(e ->  Mono.empty())
	.doFinally(__ -> logger.info(snowflake.asString())))
.subscribe();
```
Here is the execution trace I get on `3.0.1`: https://pastebin.com/wUz4uSE1
Here is the execution trace I get on `474fcee`: https://pastebin.com/2H3SVdyt

As you can see, I don't experience the disconnecting issue with the bugfix commit, however the execution is much slower and basically there is no parallelism at all. Each request is done sequentially and one by one, exactly like if I did `concatMap` with the old version.

Actually, it's even worse than a `concatMap`, because in fact it prevents any parallellism **across the entire app**. This means that even if you have two distinct Reactor chains doing their own REST requests, like so:

```Java
// Flux A
Flux.<Snowflake>create(emitter -> {
    for (long l = 1 ; l < 100 ; l++) // Note that I reduced to 100 requests for this example
        emitter.next(Snowflake.of(l));
}).flatMap(snowflake -> discordClient.getChannelById(snowflake)
        .onErrorResume(e -> Mono.empty())
        .doFinally(__ -> logger.info("Flux A: " + snowflake.asString())))
.subscribe();
// Just to ensure that the flatMap of Flux A has already finished subscribing to all REST requests
Mono.delay(Duration.ofMillis(50)).block();
// Flux B
Flux.<Snowflake>create(emitter -> {
    for (long l = 1 ; l < 100 ; l++)
        emitter.next(Snowflake.of(l));
}).flatMap(snowflake -> discordClient.getChannelById(snowflake)
        .onErrorResume(e -> Mono.empty())
        .doFinally(__ -> logger.info("Flux B: " + snowflake.asString())))
.subscribe();
```
It will result in the following output:
```
21:01:55.756 [elastic-4] INFO  bot - Flux A: 1
21:01:55.888 [elastic-4] INFO  bot - Flux A: 2
21:01:56.008 [elastic-4] INFO  bot - Flux A: 3
21:01:56.210 [elastic-4] INFO  bot - Flux A: 4
21:01:56.354 [elastic-4] INFO  bot - Flux A: 5
21:01:56.475 [elastic-4] INFO  bot - Flux A: 6
21:01:56.606 [elastic-4] INFO  bot - Flux A: ...
21:01:56.730 [elastic-4] INFO  bot - Flux A: 98
21:01:56.852 [elastic-4] INFO  bot - Flux A: 99
21:01:56.990 [elastic-4] INFO  bot - Flux B: 1
21:01:57.308 [elastic-4] INFO  bot - Flux B: 2
21:01:57.452 [elastic-4] INFO  bot - Flux B: 3
21:01:57.605 [elastic-4] INFO  bot - Flux B: 4
21:01:57.744 [elastic-4] INFO  bot - Flux B: 5
21:01:57.870 [elastic-4] INFO  bot - Flux B: 6
21:01:57.997 [elastic-4] INFO  bot - Flux B: ...
21:01:58.119 [elastic-4] INFO  bot - Flux B: 98
21:01:58.254 [elastic-4] INFO  bot - Flux B: 99
```
Even though I only put a delay of 50ms between the subscription of the two Flux, the Flux B will wait for A to finish ALL of its requests before starting. And this apply even if the two Flux are in completely different parts of the app. This essentially means that two bot commands cannot interact with Discord in parallel, which is a huge problem for large bots like mine.

In conclusion, yeah the issue is "fixed", but the ability to perform REST requests in parallel has been completely removed, which is a bad thing as you can see... It's almost like doing blocking calls at this point.

**Comment by quanticc** [comment]
@Alex1304 Can you try with `a718f8c`. I'm experimenting to find some middle-ground.

Thanks

**Comment by Alex1304** [comment]
It works better, now at least I can do requests in parallel. However I still feel like it can be pushed even furtherI have measured the speed at which the bot can make requests, using the following code:
```java
Flux.range(0, 1000)
		.map(Snowflake::of)
		.flatMap(sn -> discordClient.getChannelById(sn)
				.onErrorResume(e -> Mono.empty()))
		.then(Mono.just(0))
		.elapsed()
		.map(TupleUtils.function((time, __) -> time / 1000))
		.doOnNext(speed -> logger.info("Speed: {} requests/s", speed))
		.then();
```
and with this I get an average of 20 requests/s, which I think is much lower than the global rate limit.

I looked at the GlobalRateLimiter code just by curiosity, does the line `private final Semaphore outer = new Semaphore(8, true);` means that you are allowing a maximum of 8 parallel requests? If so, maybe you can try changing this value to let's say 20?

**Comment by quanticc** [comment]
Are you hitting the global limit this way? Then it's 50 requests/s.

We are discussing if it makes sense to parameterize that value, as throughput can vary across machines.

**Comment by Alex1304** [comment]
I'll clone the project and change the value myself so I can tell you which one gives the best results for me.

**Comment by Alex1304** [comment]
Okay, so actually i was completely wong, I'm really sorry for that. With `a718f8c` I can effectively hit the global rate limit, I just made a big mistake in the code I gave to measure the speed.
I realized it when I saw that increasing the value of the Semaphore had strictly no effect. I decided to enable discord4j.rest.traces in the logger and I saw that it's actually hitting the limit.

The error in my code resided in the `.map(TupleUtils.function((time, __) -> time / 1000))`, actually this just gives the number of *seconds*, not the number of *requests per second*. I had to change it to `1000 / (time / 1000)` in order to get the value I wanted, and now indeed it gives me a result close to 50.

I should've paid attention to this stupid mistake earlier, sorry again lol. I think this issue is definitely fixed then. Just to make sure, I have re-tested this code with 100_000 requests, and the disconnection issue never occurs.
